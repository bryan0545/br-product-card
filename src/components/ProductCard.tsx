import styles from '../styles/styles.module.css';
import { useProduct } from '../hooks/useProduct';
import React, { createContext, CSSProperties } from 'react';
import {
  Product,
  ProductContextProps,
  onChangeArgs,
  InitialValues,
  ProductCartHandlers,
} from '../interfaces/interfaces';
import { ReactElement } from 'react';

export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children: (args: ProductCartHandlers) => ReactElement;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    counter,
    increseCounter,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider value={{ counter, increseCounter, product, maxCount }}>
      <div style={style} className={`${styles.productCard} ${className}`}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          increseCounter,
          reset,
        })}
      </div>
    </Provider>
  );
};
