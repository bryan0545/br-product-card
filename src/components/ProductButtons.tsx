import React, { CSSProperties, useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

interface Props {
  className?: string;
  style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {
  const { counter, increseCounter, maxCount } = useContext(ProductContext);
  const isMaxReached = useCallback(() => !!maxCount && counter === maxCount, [
    counter,
    maxCount,
  ]);

  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button onClick={() => increseCounter(-1)} className={styles.buttonMinus}>
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        onClick={() => increseCounter(1)}
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
      >
        +
      </button>
    </div>
  );
};
