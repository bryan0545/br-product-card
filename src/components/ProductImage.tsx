import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

interface Props {
  img?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ className, img, style }: Props) => {
  const { product } = useContext(ProductContext);
  let imageToShow: string;

  if (img) {
    imageToShow = img;
  } else if (product.img) {
    imageToShow = product.img;
  } else {
    imageToShow = noImage;
  }
  return (
    <img
      style={style}
      className={`${styles.productImg} ${className}`}
      src={imageToShow}
      alt="Product"
    />
  );
};
