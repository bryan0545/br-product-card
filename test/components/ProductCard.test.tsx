import * as React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components';
import { product2 } from '../data/products';

describe('ProductCard', () => {
  const { act } = renderer;
  test('Should display the component correctly', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <h1>Hola</h1>}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('Should increase the counter', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {({ count, increseCounter }) => (
          <>
            <h1>Hola</h1>
            <span>{count}</span>
            <button onClick={() => increseCounter(1)}>+1</button>
          </>
        )}
      </ProductCard>
    );
    let tree = wrapper.toJSON();

    act(() => {
      (tree as any).children[2].props.onClick();
    });

    tree = wrapper.toJSON();

    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
