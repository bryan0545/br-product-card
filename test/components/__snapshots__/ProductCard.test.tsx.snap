// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Should display the component correctly 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Hola
  </h1>
</div>
`;
